designs/v1beta1-api.md

```diff
@@ -0,0 +1,341 @@
+     ephemeral-storage: "100Gi"
+```
+
+### `compute.k8s.aws/NodeClass`
+
+```
+apiVersion: compute.k8s.aws/v1beta1
+kind: NodeClass
+metadata:
+  name: default
+spec:
+  amiFamily: AL2
+  amiSelectorTerms:
+    - tags: 
+        key: value
+    - id: abc-123
+    - name: foo
+      owner: amazon
+    - ssm: "/my/custom/parameter"
+  subnetSelectorTerms:
+    - tags:
+        compute.k8s.aws/discovery: cluster-name
+    - id: subnet-1234
+  securityGroupSelectorTerms:
+    - tags:
+        compute.k8s.aws/discovery: cluster-name
+    - name: default-security-group
+  role: karpenter-node-role
+  userData: |
+    echo "this is custom user data"
+  tags:
+    custom-tag: custom-value
+  metadataOptions:
+    httpEndpoint: enabled
+    httpProtocolIPv6: disabled
+    httpPutResponseHopLimit: 2
+    httpTokens: required
+  blockDeviceMappings:
+    - deviceName: /dev/xvda
+      ebs:
+        volumeSize: 20Gi
+        volumeType: gp3
+        encrypted: true
+  detailedMonitoring: true
+status:
+  subnets:
+    - id: subnet-0a462d98193ff9fac
+      zone: us-east-2b
+    - id: subnet-0322dfafd76a609b6
+      zone: us-east-2c
+    - id: subnet-0727ef01daf4ac9fe
+      zone: us-east-2b
+    - id: subnet-00c99aeafe2a70304
+      zone: us-east-2a
+    - id: subnet-023b232fd5eb0028e
+      zone: us-east-2c
+    - id: subnet-03941e7ad6afeaa72
+      zone: us-east-2a
+  securityGroups:
+    - id: sg-041513b454818610b
+      name: ClusterSharedNodeSecurityGroup
+    - id: sg-0286715698b894bca
+      name: ControlPlaneSecurityGroup-1AQ073TSAAPW
+  amis:
+    - id: ami-05a05e85b17bb60d7
+      name: amazon-eks-node-1.24-v20230703
+      requirements:
+        - key: karpenter.k8s.aws/instance-accelerator-count
+          operator: DoesNotExist
+        - key: kubernetes.io/arch
+          operator: In
+          values:
+            - amd64
+        - key: karpenter.k8s.aws/instance-gpu-count
+          operator: DoesNotExist
+    - id: ami-0d849ef1e65103147
+      name: amazon-eks-gpu-node-1.24-v20230703
+      requirements:
+        - key: kubernetes.io/arch
+          operator: In
+          values:
+            - amd64
+        - key: karpenter.k8s.aws/instance-accelerator-count
+          operator: Exists
+    - id: ami-0d849ef1e65103147
+      name: amazon-eks-gpu-node-1.24-v20230703
+      requirements:
+        - key: kubernetes.io/arch
+          operator: In
+          values:
+           - amd64
+        - key: karpenter.k8s.aws/instance-gpu-count
+          operator: Exists
+    - id: ami-0c3487f30d003deb3
+      name: amazon-eks-arm64-node-1.24-v20230703
+      requirements:
+        - key: kubernetes.io/arch
+          operator: In
+          values:
+            - arm64
+        - key: karpenter.k8s.aws/instance-gpu-count
+          operator: DoesNotExist
+        - key: karpenter.k8s.aws/instance-accelerator-count
+          operator: DoesNotExist
+```
+
+### `karpenter.sh/NodeClaim`
+
+```
+apiVersion: karpenter.sh/v1beta1
```

Comment:
Why is NodePool and NodeClaim `karpenter.sh/v1beta1` but NodeClass is `compute.k8s.aws/v1beta1`?

Reply:
We don't want to be managing the CloudProvider-specific CRDs in our top-level API Group, since according to upstream guidance, this means that they would have to be versioned and managed together. 

Guidance from upstream and also from @sftim was that these should be split out into their own cloudprovider-specific API groups so that each cloudprovider is free to manage the CRDs and version them as they see fit.

Reply:
Is there a way for users to validate the support matrix of karpenter, NodeClass, and their cloud provider's APIs?

Reply:
> Is there a way for users to validate the support matrix of karpenter, NodeClass, and their cloud provider's APIs?

In general? It would depend on the provider. The design here anticipates having more than once cloud provider integrations.

(Personally, I would call the AWS node class kind an AWSNodeClass, in case someone is running a hybrid cluster with eg a mix of AWSNodeClass and ExampleBareMetalNode objects)

Reply:
One option I discussed with Jonathan is to call it EC2NodeClass but make a shortname for nodeclass. Alternatively, you could make the CRD NodeClass, and make the shortname ec2nodeclass. 

Ideally, we could just let kubectl support multiple objects with the same name but different group. Would be useful for Knative Services, etc.

Reply:
> Ideally, we could just let kubectl support multiple objects with the same name but different group. Would be useful for Knative Services, etc.

Sure, but: do we want to hold off v1beta1 until that's done? I think not.

I like the idea of `ec2nodeclass` and also `ec2nc` as a short name. The Helm chart could plausibly let you decide on whether to make just `nodeclass` also be a short name (defaults to off; AWS docs would encourage flipping that to on).

Reply:
Anyway, +1 for EC2NodeClass.

Reply:
AFAIK, you can't template CRDs in helm charts so I think it would have to be on by default.

Reply:
Ah yep.

Reply:
> you can't template CRDs in helm charts so I think it would have to be on by default

Because of the concern for conflicts right now, we're going to punt on short-naming the `EC2NodeClass` with `nodeclass`. We can always support this down-the-line and it might be nice for `kubectl`/apiserver to support multiple objects with the same shortname and to print them in a way that is friendly for that.